src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -2212,53 +2212,66 @@ String StorageReplicatedMergeTree::getChecksumsForZooKeeper(const MergeTreeDataP
-            const auto part_info = MergeTreePartInfo::tryParsePartName(it->name(), format_version);
+            part->loadChecksums(true);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log,
+                fmt::format("data race is possible when reading checksums.txt from detached part or part does not have a file checksums.txt, part {} is ignored", detached_part_info.dir_name));
+            return true;
+        }
 
-            if (!part_info || part_info->getPartitionId() != actual_part_info.getPartitionId())
-                continue;
+        return entry.part_checksum != part->checksums.getTotalChecksumHex();
+    });
 
-            const auto part_old_name = part_info->getPartNameV1();
-            const auto volume = std::make_shared<SingleDiskVolume>("volume_" + part_old_name, disk);
+    if (detached_parts.empty())
+        return {};
 
-            auto part = getDataPartBuilder(entry.new_part_name, volume, fs::path(DETACHED_DIR_NAME) / part_old_name, getReadSettings())
-                .withPartFormatFromDisk()
-                .build();
+    auto & detached_part_info = detached_parts.front();
 
-            try
-            {
-                part->loadColumnsChecksumsIndexes(
-                    /* require_columns_checksums = */ true, /* check_consistency = */ true, /* load_metadata_version = */ false);
-            }
-            catch (const Exception&)
-            {
-                /// This method throws if the part data is corrupted or partly missing. In this case, we simply don't
-                /// process the part.
-                continue;
-            }
+    rename_parts.addPart(detached_part_info.dir_name, "attaching_" + detached_part_info.dir_name, detached_part_info.disk);
 
-            if (entry.part_checksum == part->checksums.getTotalChecksumHex())
-            {
-                auto metadata_version = getInMemoryMetadataPtr()->getMetadataVersion();
-                part->writeMetadataVersion(getContext(), metadata_version, (*getSettings())[MergeTreeSetting::fsync_after_insert]);
-                part->setMetadataVersion(metadata_version);
-                part->modification_time = part->getDataPartStorage().getLastModified().epochTime();
-                return part;
-            }
-        }
+    try
+    {
+        rename_parts.tryRenameAll();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(log, fmt::format("data race is possible when moving detached part, part {} is ignored", detached_part_info.dir_name));
+        return {};
     }
 
-    return {};
+    const auto volume = std::make_shared<SingleDiskVolume>("volume_" + detached_part_info.dir_name, detached_part_info.disk);
+    auto part = getDataPartBuilder(entry.new_part_name, volume, fs::path(rename_parts.source_dir) / rename_parts.old_and_new_names.front().new_name, getReadSettings())
+        .withPartFormatFromDisk()
+        .build();
+
+    loadPartAndFixMetadataImpl(part, getContext());
```

Comment:
If part was corrupted and had empty prefix -- this can happen in case of data corruption after detach then we will not be able to execute AttachPart log entry at all, because `loadPartAndFixMetadataImpl` will validate checksums and throw an exception which will skip download of part from other replicas and probably just postpone the execution of entry

Reply:
Cool, that is the case in `test_attach_without_fetching/test.py::test_attach_without_fetching`.